File: leetcode/199-binary-tree-right-side-view/main.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// dfs, iterative
type Stack struct {
	Node  *TreeNode
	Route []int
}

func rightSideView(root *TreeNode) []int {
	result := []int{}
	if root == nil {
		return result
	}
	stack := []*Stack{}
	stack = append(stack, &Stack{root, []int{root.Val}})
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		node := pop.Node
		route := pop.Route
		// put left child to the stack first, since we want to read right child in the next iteration
		if node.Left == nil && node.Right == nil && len(route) > len(result) {
			result = append(result, route[len(result):]...)
		}
		if node.Left != nil {
			stack = append(stack, &Stack{node.Left, append(route, node.Left.Val)})
		}
		if node.Right != nil {
			stack = append(stack, &Stack{node.Right, append(route, node.Right.Val)})
		}
	}
	return result
}

// recursive
func rightSideView1(root *TreeNode) []int {
	result := []int{}
	if root == nil {
		return result
	}
	var dfs func(root *TreeNode, route []int)
	dfs = func(root *TreeNode, route []int) {
		if root != nil {
			if root.Left == nil && root.Right == nil && len(route) > len(result) {
				result = append(result, route[len(result):]...)
			}
			// put right child first, since we want the right first
			if root.Right != nil {
				dfs(root.Right, append(route, root.Right.Val))
			}
			if root.Left != nil {
				dfs(root.Left, append(route, root.Left.Val))
			}
		}
	}
	dfs(root, []int{root.Val})
	return result
}

func main() {
	// 		5
	//	2		8
	// 1 3 7
	//10
	root := &TreeNode{5,
		&TreeNode{2,
			&TreeNode{1,
				&TreeNode{10, nil, nil},
				nil,
			},
			&TreeNode{3, nil, nil},
		},
		&TreeNode{8,
			&TreeNode{7, nil, nil},
			nil,
		},
	}
	ans := rightSideView(root)
	fmt.Println(ans)
	ans = rightSideView1(root)
	fmt.Println(ans)
}

File: leetcode/199-binary-tree-right-side-view/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


"""
    1st approach: iterative dfs
    - declare a result array
    - when we traverse down, put the val into the result if the depth > len(array)

    Time    O(n)
    Space   O(h)
    24 ms, faster than 74.45%
"""


class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        res = []

        def dfs(node, depth):
            if node == None:
                return
            if depth == len(res):
                res.append(node.val)
            if node.right:
                dfs(node.right, depth + 1)
            if node.left:
                dfs(node.left, depth + 1)
        dfs(root, 0)

        return res


"""
    1st approach: bfs
    - declare a result array
    - when we traverse down, put the val into the result if the depth > len(array)

    Time    O(n)
    Space   O(h)
    16 ms, faster than 99.24%
    15may2019
"""


class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        res = []
        q = [(root, 0)]
        while len(q) > 0:
            n = len(q)
            for _ in range(n):
                node, depth = q.pop(0)
                if depth == len(res):
                    res.append(node.val)
                if node.right:
                    q.append((node.right, depth + 1))
                if node.left:
                    q.append((node.left, depth + 1))
        return res
